external/vcm/vcm/limit.py

```diff
@@ -0,0 +1,113 @@
+        limits: variable containing limits of the data, which must include a 'bounds'
+            dimension containing 'upper' and 'lower' coordinates. Other dimensions
+            should be a subset of the data dimensions.
+            
+    Returns: a limited data array
+    
+    """
+    upper = limits.sel(bounds="upper")
+    lower = limits.sel(bounds="lower")
+    return data.where(data < upper, upper).where(data > lower, lower)
+
+
+class DatasetQuantileLimiter(BaseEstimator, TransformerMixin):
+    """Transformer to reduce the extremity of outliers of a dataset to quantile limits
+    
+    Limiting is done on a variable by variable basis and along specified dimensions.
+        Limits are optionally computed on a configurable subset of the dataset to avoid
+        loading the entire dataset.
+    
+    Args:
+        alpha: two-tailed alpha for computing extrema quantiles, values beyond which
+            will be reduced to the quantile
+        limit_only: variables in the dataset to be limited; if not specified all
+            variables are limited
+
+    """
+
+    def __init__(
+        self, alpha: float, limit_only: Optional[Sequence[str]] = None,
+    ):
+        self._alpha: float = alpha
+        self._limit_only: Optional[Sequence[str]] = limit_only
+        self._limits: xr.Dataset = None
+
+    def fit(
+        self,
+        ds: xr.Dataset,
+        feature_dims: Optional[Sequence[str]] = None,
+        fit_indexers: Optional[Mapping[str, int]] = None,
+    ) -> "DatasetQuantileLimiter":
+        """Fit the limiter on a dataset.
+        
+        Args:
+            ds: Dataset to be used to fit the limits.
+            feature_dims: Dimensions along which quantile limits should NOT be
+                computed, i.e., the resulting limits will be computed separately
+                for each coordinate along the feature_dims.
+            fit_indexer: Integer-based indexers that select a subset of `ds` on
+                which to fit limits. Indexed dimensions must not be in `feature_dims`.
+            
+        Returns: Fitted limiter
+        
+        """
+        if fit_indexers is not None and feature_dims is not None:
+            for index_dim in fit_indexers.keys():
+                if index_dim in feature_dims:
+                    raise ValueError(
+                        f"Indexer dim {index_dim} may not be in feature_dims."
+                    )
+
+        sample_ds = (ds.isel(**fit_indexers) if fit_indexers is not None else ds).load()
+        sample_dims = (
+            set(sample_ds.dims) - set(feature_dims)
+            if feature_dims is not None
+            else sample_ds.dims
+        )
+        lower = sample_ds.quantile(self._alpha / 2.0, dim=sample_dims)
+        upper = sample_ds.quantile(1.0 - self._alpha / 2.0, dim=sample_dims)
+        self._limits = xr.concat(
+            [lower, upper],
+            dim=xr.DataArray(["lower", "upper"], dims=["bounds"], name="bounds"),
+        )
```

Comment:
Maybe simpler to just store the upper and lower bounds in different attrs?
```suggestion
        self._lower_limit = sample_ds.quantile(self._alpha / 2.0, dim=sample_dims)
        self._upper_limit = sample_ds.quantile(1.0 - self._alpha / 2.0, dim=sample_dims)
```